internal/server/server_authz_checks.go

```diff
@@ -2,16 +2,66 @@ package server
 
 import (
 	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/subtle"
+	"encoding/hex"
+	"io"
 	"net/http"
 
 	"github.com/kopia/kopia/internal/auth"
 )
 
+const kopiaSessionCookie = "Kopia-Session-Cookie"
+
+func (s *Server) generateCSRFToken(sessionID string) string {
+	h := hmac.New(sha256.New, s.authCookieSigningKey)
+	io.WriteString(h, sessionID) //nolint:errcheck
+
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func (s *Server) validateCSRFToken(r *http.Request) bool {
+	if s.options.DisableCSRFTokenChecks {
+		return true
+	}
+
+	ctx := r.Context()
+	path := r.URL.Path
+
+	sessionCookie, err := r.Cookie(kopiaSessionCookie)
+	if err != nil {
+		log(ctx).Warnf("missing or invalid session cookie for %v", path)
```

Comment:
Best to also include the error in the log.

```suggestion
		log(ctx).Warnf("missing or invalid session cookie for %v: %v", path, err)
```

Reply:
done